### PR TITLE
[MAINTENANCE] Start refactoring codebase to use checkpoint factory CRUD

### PR DESCRIPTION
**Pull request by cdkini**
- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/checkpoint/checkpoint.py [hunk 1]
```diff
@@ -102,7 +102,7 @@ class BaseCheckpoint(ConfigPeer):
     def __init__(
         self,
         checkpoint_config: CheckpointConfig,
-        data_context: AbstractDataContext,
+        data_context: AbstractDataContext | None = None,
```
Comment (cdkini) [inline great_expectations/checkpoint/checkpoint.py]: I'd like to fully decouple this requirement but for now, let's make it optional and have the factory supply it on `add`

File: great_expectations/checkpoint/checkpoint.py [hunk 2]
```diff
@@ -225,8 +232,8 @@ def run(  # noqa: C901, PLR0913, PLR0915
             "expectation_suite_ge_cloud_id": expectation_suite_ge_cloud_id,
         }
 
-        substituted_runtime_config: dict = self.get_substituted_config(
-            runtime_kwargs=runtime_kwargs
+        substituted_runtime_config: dict = self._get_substituted_config(
+            config_provider=context.config_provider, runtime_kwargs=runtime_kwargs
```
Comment (cdkini) [inline great_expectations/checkpoint/checkpoint.py]: let's make this private and pass explicit context dependencies down so we only have to do a single assert at the top level of this func

File: great_expectations/checkpoint/checkpoint.py [hunk 3]
```diff
@@ -318,40 +327,24 @@ def run(  # noqa: C901, PLR0913, PLR0915
...
-        return substituted_runtime_config
-
-    def _get_substituted_runtime_kwargs(
-        self,
-        source_config: dict,
-        runtime_kwargs: dict | None = None,
-    ) -> dict:
-        if runtime_kwargs is None:
-            runtime_kwargs = {}
-
```
Comment (cdkini) [inline great_expectations/checkpoint/checkpoint.py]: Three helper funcs that can easily be compressed into one

File: great_expectations/data_context/types/base.py [hunk 4]
```diff
@@ -2825,110 +2821,6 @@ def __str__(self) -> str:
         """
         return self.__repr__()
 
-    # noinspection PyUnusedLocal,PyUnresolvedReferences
-    @staticmethod
-    def resolve_config_using_acceptable_arguments(  # noqa: PLR0913
```
Comment (cdkini) [inline great_expectations/data_context/types/base.py]: Not used anywhere so let's delete

File: docs/docusaurus/docs/cloud/connect/connect_airflow.md [hunk 5]
```diff
@@ -74,7 +74,7 @@ Apache Airflow is an orchestration tool that allows you to schedule and monitor
     def run_gx_airflow():
 
         context = gx.get_context()
-        checkpoint = context.get_checkpoint(name = CHECKPOINT_NAME)
+        checkpoint = context.checkpoints.get(name = CHECKPOINT_NAME)
```
Comment (tyler-hoffman) [inline docs/docusaurus/docs/cloud/connect/connect_airflow.md:77]: totally unnecessary to change, but are we not autoformatting docs? the spaces around the `=` look so foreign to me!

File: great_expectations/data_context/store/configuration_store.py [hunk 6]
```diff
@@ -112,6 +113,10 @@ def deserialize(self, value):
         except gx_exceptions.InvalidBaseYamlConfigError:
             # Just to be explicit about what we intended to catch
             raise
+        except marshmallow.ValidationError as e:
```
Comment (tyler-hoffman) [inline great_expectations/data_context/store/configuration_store.py:116]: Still looking, but I wouldn't have expected this change. Why is this coming up in this PR?

**Review by tyler-hoffman — APPROVED** [review]
Wondering about the marshmellow error, but LGTM

**Pull request merged**